File: java-algo/src/main/java/com/leetcode/Problem808.java
package com.leetcode;

/**
 * 808. Soup Servings
 * https://leetcode.com/problems/soup-servings/
 */
public class Problem808 {
    public double soupServings(int n) {
        n = (int) Math.ceil((double) n / 25);
        if (n >= 179) {
            return 1;
        }
        double[][] memo = new double[n + 1][n + 1];
        return dfs(n, n, memo);
    }

    private double dfs(int x, int y, double[][] memo) {
        if (x <= 0 && y <= 0) {
            return 0.5;
        }
        if (x <= 0) {
            return 1;
        }
        if (y <= 0) {
            return 0;
        }
        if (memo[x][y] != 0) {
            return memo[x][y];
        }
        return memo[x][y] = 0.25 * (dfs(x - 4, y, memo) + dfs(x - 3, y - 1, memo) + dfs(x - 2, y - 2, memo) + dfs(x - 1, y - 3, memo));
    }
}

File: python-algo/leetcode/problem_808.py
"""808. Soup Servings
https://leetcode.com/problems/soup-servings/
"""
import math


class Solution:
    def soupServings(self, n: int) -> float:
        n = math.ceil(n / 25)
        if n >= 200:
            return 1
        dp = [[0] * (n + 1) for _ in range(n + 1)]
        dp[0][0] = 0.5
        for i in range(1, n + 1):
            dp[0][i] = 1
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                a = dp[max(i - 4, 0)][j]
                b = dp[max(i - 3, 0)][max(j - 1, 0)]
                c = dp[max(i - 2, 0)][max(j - 2, 0)]
                d = dp[max(i - 1, 0)][max(j - 3, 0)]
                dp[i][j] = 0.25 * (a + b + c + d)
        return dp[n][n]
